### PR TITLE
Stratify sampling when split train/test data

### PR DESCRIPTION
**Pull request by YuxinB**
Fixes #

Changes proposed in this pull request:

-

## Before submitting

- [ ] I've read and followed all steps in the [Making a pull request](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/py-why/pywhy-graphs/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.

### CODE REVIEW

File: sktree/stats/forestht.py
Comment (adam2392) [inline sktree/stats/forestht.py]: If y will be passed. There should be a test that y is of the same type during fitting. 

I.e. type of target, n samples, n outputs should be invariant.

**Review by adam2392 — CHANGES_REQUESTED** [review]
If public API changes we need a test to control accidental misuse

**Review by PSSF23 — CHANGES_REQUESTED** [review]
The changes I marked are likely regressions and should be removed.

File: sktree/stats/forestht.py
Comment (PSSF23) [inline sktree/stats/forestht.py:468]: These changes are regressions and should be removed

File: sktree/stats/forestht.py
Comment (PSSF23) [inline sktree/stats/forestht.py]: These changes are regressions and should be removed

File: sktree/stats/forestht.py
Comment (PSSF23) [inline sktree/stats/forestht.py]: These changes are regressions and should be removed

File: sktree/stats/forestht.py
Comment (YuxinB) [reply sktree/stats/forestht.py]: Okay, adding a check type now.

**Review by PSSF23 — COMMENTED** [review]
I have corrected the change regressions & formatted the file.

**Review by PSSF23 — COMMENTED** [review]
@adam2392 @YuxinB The failed checks are about `train_test_samples_`, which takes the argument of stratifier y and not recognized as an attribute in the test. The `property` label was removed. How do you think this should be resolved?

**Comment by adam2392** [comment]
> @adam2392 @YuxinB The failed checks are about `train_test_samples_`, which takes the argument of stratifier y and not recognized as an attribute in the test. The `property` label was removed. How do you think this should be resolved?

I think the issue is that `train_test_samples_` is meant to be a "fitted" property that can be recomputed after-the-fact. However, a reliance on `y` essentially means it is not replicable without access to `y`.

There are a few ideas that come to mind:

- Add `_y` as a cached array: this increases RAM usage and disc usage when picking, but removes any reliance on keeping track of `y`
- Only pass `y` into a private property
- Rename `train_test_samples_` and add a function outside the class `_compute_train_test_samples`, so we don't expose what are the training/testing indices

I am in favor of the first option which is only activated when we need `y` to be stratified (i.e. classification)

**Comment by adam2392** [comment]
I've implemented said changes. @YuxinB you must add yourself to the `contributors.rst` file. You also need to test an additional two if/else statement blocks.

Assuming CIs work, I'll let @sampan501 and @PSSF23 review and merge if they are happy.

File: examples/hypothesis_testing/plot_MI_gigantic_hypothesis_testing_forest.py
Comment (adam2392) [inline examples/hypothesis_testing/plot_MI_gigantic_hypothesis_testing_forest.py:1]: Made example simpler

File: sktree/stats/forestht.py
Comment (adam2392) [inline sktree/stats/forestht.py]: Please add a test for these lines.

File: sktree/stats/forestht.py
Comment (adam2392) [inline sktree/stats/forestht.py]: Please add a test for these lines.

**Review by sampan501 — DISMISSED** [review]
I think the other thing that is missing is documentation for the `stratify` parameter

**Review by PSSF23 — COMMENTED** [review]
Previously the tests for stratification were not working (they will always pass). I think I corrected them & added new ones @adam2392 requested.

**Review by PSSF23 — COMMENTED** [review]
EDIT: I redid the tests, as the previous test format can never test stratification.

**Review by PSSF23 — COMMENTED** [review]
I removed duplicate checks as those would be covered by `check_input` anyway. Explicitly setting the parameter to `False` just to have new checks that check the same thing doesn't make sense to me.

**Review by PSSF23 — APPROVED** [review]
@adam2392 I believe the tests should be excluded from codecov? And do you think we need more tests?

**Comment by adam2392** [comment]
Can you add docstring for stratify in the two classes?

**Review by PSSF23 — COMMENTED** [review]
I set the base forest default value to `False` and removed it from the regressor class init

**Comment by sampan501** [comment]
What set the default of `stratify` to false? Isn't it better to stratify in general according to the sims? Or is it because of the variance in the performance?

**Comment by adam2392** [comment]
Yeah just to give options.

**Comment by PSSF23** [comment]
> What set the default of `stratify` to false? Isn't it better to stratify in general according to the sims? Or is it because of the variance in the performance?

I only set the base forest value to false for regressor, classifier is still default true

**Comment by PSSF23** [comment]
@adam2392 Do you know how to exclude `experimental` and all tests from codecov?

**Comment by adam2392** [comment]
> @adam2392 Do you know how to exclude `experimental` and all tests from codecov?

Experimental we can include, but the tests should be not part of it. I think this might be a bug on codecov cuz it wasn't showing up before.

**Comment by adam2392** [comment]
The build-docs are failing because @YuxinB needs to add herself to the contributors doc: 

```
/home/circleci/project/doc/whats_new/v0.3.rst:18: ERROR: Unknown target name: "yuxin bai".
/home/circleci/project/doc/whats_new/v0.3.rst:28: ERROR: Unknown target name: "yuxin bai".
```

The codecov/project doesn't need to pass

**Review by adam2392 — APPROVED** [review]
LGTM once CI green

**Comment by adam2392** [comment]
You'll need to rename the references to the example file, since you changed the name of the example.

```
/home/circleci/project/doc/auto_examples/hypothesis_testing/plot_MI_imbalanced_hyppo_testing.rst:39: WARNING: undefined label: 'sphx_glr_auto_examples_hypothesis_testing_plot_mi_gigantic_hypothesis_testing_forest.py'
```

**Comment by PSSF23** [comment]
> You'll need to rename the references to the example file, since you changed the name of the example.
> 
> ```
> /home/circleci/project/doc/auto_examples/hypothesis_testing/plot_MI_imbalanced_hyppo_testing.rst:39: WARNING: undefined label: 'sphx_glr_auto_examples_hypothesis_testing_plot_mi_gigantic_hypothesis_testing_forest.py'
> ```

Done

**Review by adam2392 — APPROVED** [review]

**Comment by adam2392** [comment]
Thanks @YuxinB @PSSF23 and @sampan501

**Pull request merged**
